src/SFA.DAS.TokenService.Application/PrivilegedAccess/TokenRefresh/HmrcAuthTokenBroker.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using NLog;
+using SFA.DAS.TokenService.Domain;
+using SFA.DAS.TokenService.Domain.Data;
+using SFA.DAS.TokenService.Domain.Services;
+using SFA.DAS.TokenService.Infrastructure.ExecutionPolicies;
+
+namespace SFA.DAS.TokenService.Application.PrivilegedAccess.TokenRefresh
+{
+    public sealed class HmrcAuthTokenBroker : IHmrcAuthTokenBroker, IDisposable
+    {
+        private const string PrivilegedAccessSecretName = "PrivilegedAccessSecret";
+
+        private readonly ExecutionPolicy _executionPolicy;
+        private readonly ILogger _logger;
+        private readonly IOAuthTokenService _tokenService;
+        private readonly ISecretRepository _secretRepository;
+        private readonly ITotpService _totpService;
+        private readonly ITokenRefresher _tokenRefresher;
+
+        private readonly Task<OAuthAccessToken> _initialiseTask;
+        private OAuthAccessToken _cachedAccessToken;
+
+        private CancellationTokenSource _cancellationTokenSource;
+
+        public HmrcAuthTokenBroker(
+            [RequiredPolicy(HmrcExecutionPolicy.Name)] ExecutionPolicy executionPolicy, 
+            ILogger logger,
+            IOAuthTokenService tokenService,
+            ISecretRepository secretRepository,
+            ITotpService totpService,
+            ITokenRefresher tokenRefresher)
+        {
+            _secretRepository = secretRepository;
+            _totpService = totpService;
+            _tokenService = tokenService;
+            _logger = logger;
+            _executionPolicy = executionPolicy;
+            _tokenRefresher = tokenRefresher;
+            _initialiseTask = InitialiseToken();
```

Comment:
This will be called multiple times due to a bug in StructureMap prior to version 4.6 (https://github.com/structuremap/structuremap/pull/625). I recommend upgrading to the latest version or avoid any expensive calls in singleton constructors.

Reply:
'tis done